### DISCUSSION TITLE
uPickle 4.0.0-RC1

### DISCUSSION THREAD

**Discussion by lihaoyi** [lihaoyi/upickle#609] [Announcements]
I just cut uPickle 4.0.0-RC1 https://com-lihaoyi.github.io/upickle/#4.0.0-RC1

uPickle 4.0.0 is a major, breaking change. It fixes a few of the oldest warts and inconveniences of the uPickle library, in particular:

* Shorten $type tag used for discriminating sealed trait cases [#594](https://github.com/com-lihaoyi/upickle/pull/596)
* Unbox first level of Options when serializing JSON [#598](https://github.com/com-lihaoyi/upickle/pull/598)

This is at a cost of breaking both binary compatibility and *serialization compatibility*. JSON serialized by uPickle 2.x or 3.x is *not* deserializable by uPickle 4.x, unless you follow the steps described in the changelog.

4.0.0-RC1 is meant for people to try out before we cut a version 4.0.0 final. Please go through the changelog, give it a spin and let me know if you find any issues. uPickle does not make breaking changes often (last bincompat breakage was mid 2023, last serialization breakage was mid-2022), so once 4.0.0 is finalized I hope to not need to anymore breaking changes for a while to come, and having people try out 4.0.0-RC1 would be a great help in achieving that

If no major issues are reported, I intend to cut uPickle 4.0.0 final by the end of July 2024

**Comment by plokhotnyuk** [comment]
Will uPickle 4.0.0 be released for 3.3 LTS?

**Reply by sideeffffect** [reply]
* https://github.com/com-lihaoyi/upickle/pull/610

**Reply by lihaoyi** [reply]
We hit compiler crashes that were resolved in 3.4.2, so the answer would depend on if anyone is able to fix them. They should repro in the linked PR

**Reply by sideeffffect** [reply]
I see... :( oh well
Thanks for letting me know.

**Comment by lihaoyi** [comment]
uPickle 4.0.0 final is out
